src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs

```diff
@@ -864,6 +865,31 @@ public ErrorDiagnostic ArgumentCountMismatch(int argumentCount, int mininumArgum
                 TextSpan,
                 "BCP152",
                 $"Function \"{functionName}\" cannot be used as a decorator.");
+       
+            public ErrorDiagnostic InvalidResourceTypeSegment(string typeSegment) => new(
+                TextSpan,
+                "BCP153",
+                $"The resource type segment \"{typeSegment}\" is invalid. Nested resources must specify a single type segment, and optionally can specify an api version using the format \"<type>@<apiVersion>\".");
+
+            public ErrorDiagnostic InvalidAncestorResourceType(string resourceName) => new(
+                TextSpan,
+                "BCP154",
+                $"The resource type cannot be determined due to an error in containing resource \"{resourceName}\".");
+
+            public ErrorDiagnostic ResourceRequiredForResourceAccess(string wrongType) => new(
+                TextSpan,
+                "BCP155",
+                $"Cannot access nested resources of type \"{wrongType}\". A resource type is required.");
+
+            public ErrorDiagnostic NestedResourceNotFound(string resourceName, string identifierName, IEnumerable<string> nestedResourceNames) => new(
+                TextSpan,
+                "BCP156",
+                $"The resource \"{resourceName}\" does not contain a nested resource named \"{identifierName}\". Known nested resources are: {string.Join(", ", nestedResourceNames.Select(r => $"\"{r}\""))}.");
+
+            public ErrorDiagnostic NestedResourceNotAllowedInLoop() => new(
+                TextSpan,
+                "BCP157",
+                $"Nested resource cannot appear inside of loops.");

```

Comment:
FYI, most of our other diagnostics are referring to loops as "for-expressions" (see above in this file) - would be good to be consistent with terminology.

Reply:
Happy to update. 

You might want to look through the others. I looked at some other messages for consistency and some of them also refer to loops.